# 🌎 Welcome to the CSE151B Spring 2025 Climate Emulation Competition!

Thank you for participating in this exciting challenge focused on building machine learning models to emulate complex climate systems.  
This notebook is provided as a **starter template** to help you:

- Understand how to load and preprocess the dataset  
- Construct a baseline model  
- Train and evaluate predictions using a PyTorch Lightning pipeline  
- Format your predictions for submission to the leaderboard  

You're encouraged to:
- Build on this structure or replace it entirely
- Try more advanced models and training strategies
- Incorporate your own ideas to push the boundaries of what's possible

If you're interested in developing within a repository structure and/or use helpful tools like configuration management (based on Hydra) and logging (with Weights & Biases), we recommend checking out the following Github repo. Such a structure can be useful when running multiple experiments and trying various research ideas.

👉 [https://github.com/salvaRC/cse151b-spring2025-competition](https://github.com/salvaRC/cse151b-spring2025-competition)

Good luck, have fun, and we hope you learn a lot through this process!


### 📦 Install Required Libraries
We install the necessary Python packages for data loading, deep learning, and visualization.


In [45]:
!pip install xarray zarr dask lightning matplotlib wandb cftime einops --quiet

import os
from datetime import datetime
import numpy as np
import xarray as xr
import dask.array as da
import torch
import torch.nn as nn
import torch.optim as optim
import matplotlib.pyplot as plt
from torch.utils.data import Dataset, DataLoader
import lightning.pytorch as pl


### ⚙️ Configuration Setup  
Define all model, data, and training hyperparameters in one place for easy control and reproducibility.

### 📊 Data Configuration

We define the dataset settings used for training and evaluation. This includes:

- **`path`**: Path to the `.zarr` dataset containing monthly climate variables from CMIP6 simulations.
- **`input_vars`**: Climate forcing variables (e.g., CO₂, CH₄) used as model inputs.
- **`output_vars`**: Target variables to predict — surface air temperature (`tas`) and precipitation (`pr`).
- **`target_member_id`**: Ensemble member to use from the simulations (each SSP has 3) for target variables.
- **`train_ssps`**: SSP scenarios used for training (low to high emissions).
- **`test_ssp`**: Scenario held out for evaluation (Must be set to SSP245).
- **`test_months`**: Number of months to include in the test split (Must be set to 120).
- **`batch_size`** and **`num_workers`**: Data loading parameters for PyTorch training.

These settings reflect how the challenge is structured: models must learn from some emission scenarios and generalize to unseen ones.

> ⚠️ **Important:** Do **not modify** the following test settings:
>
> - `test_ssp` must remain **`ssp245`**, which is the held-out evaluation scenario.
> - `test_months` must be **`120`**, corresponding to the last 10 years (monthly resolution) of the scenario.



In [46]:
%pwd

'/home/adapsay/teams/kaggle-group-30/CSE151B_Milestone'

In [67]:
#NOTE Change the data directory according to where you have your zarr files stored
config = {
    "data": {
        "path": "../processed_data_cse151b_v2_corrupted_ssp245/processed_data_cse151b_v2_corrupted_ssp245.zarr",
        # "path": "/kaggle/input/cse151b-spring2025-competition/processed_data_cse151b_v2_corrupted_ssp245/processed_data_cse151b_v2_corrupted_ssp245.zarr",
        "input_vars": ["CO2", "SO2", "CH4", "BC", "rsdt"],
        "output_vars": ["tas", "pr"],
        "target_member_id": 0,
        "train_ssps": ["ssp126", "ssp370", "ssp585"],
        "test_ssp": "ssp245",
        "test_months": 360,
        "batch_size": 32,
        "num_workers": 4,
        "T": 9
    },
    # "model": {
    #     "type": "simple_cnn",
    #     "kernel_size": 3,
    #     "init_dim": 64,
    #     "depth": 4,
    #     "dropout_rate": 0.1,
    # }
    # "model": {
    # "type": "convlstm",
    # "kernel_size": 3,
    # "init_dim": 64,
    # "depth": 4,
    # "dropout_rate": 0.1,
    # },
    "model": {
    "type": "convlstm_unet",  # more descriptive
    "kernel_size": 3,
    "init_dim": 64,
    "depth": 4,  # optional: remove if unused
    "dropout_rate": 0.1,
    },
    "training": {
        "lr": 1e-3,
    },
    "trainer": {
        "max_epochs": 40,
        "accelerator": "auto",
        "devices": "auto",
        "precision": 32,
        "deterministic": True,
        "num_sanity_val_steps": 0,
    },
    "seed": 42,
}
pl.seed_everything(config["seed"])  # Set seed for reproducibility

Seed set to 42


42

### 📊 Data Configuration

In [48]:
data_path = "../processed_data_cse151b_v2_corrupted_ssp245/processed_data_cse151b_v2_corrupted_ssp245.zarr"

ds = xr.open_zarr(data_path, consolidated=True)
ds

<xarray.Dataset> Size: 621MB
Dimensions:    (ssp: 4, time: 1021, latitude: 48, longitude: 72, x: 72, y: 48,
                member_id: 3)
Coordinates:
    lat        (x, y) float64 28kB dask.array<chunksize=(72, 48), meta=np.ndarray>
  * latitude   (latitude) float64 384B -89.05 -85.26 -81.47 ... 85.26 89.05
    lon        (x, y) float64 28kB dask.array<chunksize=(72, 48), meta=np.ndarray>
  * longitude  (longitude) float64 576B 1.25 6.25 11.25 ... 346.2 351.2 356.2
  * member_id  (member_id) int64 24B 0 1 2
  * ssp        (ssp) <U6 96B 'ssp126' 'ssp245' 'ssp370' 'ssp585'
  * time       (time) object 8kB 2015-01-15 00:00:00 ... 2100-01-15 00:00:00
  * x          (x) float64 576B 1.875 6.875 11.88 16.88 ... 346.9 351.9 356.9
  * y          (y) float64 384B -88.59 -84.82 -81.05 ... 81.05 84.82 88.59
Data variables:
    BC         (ssp, time, latitude, longitude) float64 113MB dask.array<chunksize=(1, 24, 48, 72), meta=np.ndarray>
    CH4        (ssp, time) float64 33kB dask.array<chunksize=(1, 24), meta=np.ndarray>
    CO2        (ssp, time) float64 33kB dask.array<chunksize=(1, 24), meta=np.ndarray>
    SO2        (ssp, time, latitude, longitude) float64 113MB dask.array<chunksize=(1, 24, 48, 72), meta=np.ndarray>
    pr         (ssp, time, member_id, y, x) float32 169MB dask.array<chunksize=(1, 24, 1, 48, 72), meta=np.ndarray>
    rsdt       (ssp, time, y, x) float32 56MB dask.array<chunksize=(1, 24, 48, 72), meta=np.ndarray>
    tas        (ssp, time, member_id, y, x) float32 169MB dask.array<chunksize=(1, 24, 1, 48, 72), meta=np.ndarray>
Attributes:
    original_member_ids:  ['r10i1p1f1', 'r11i1p1f1', 'r4i1p1f1']
    precipitation_units:  mm/day (converted from kg m-2 s-1)
    source:               CMIP6 data processed for CSE151B
    ssp:                  ssp126

### Data Exploration

**Spatial Dimensions, Size of Train/Validate/Test Data**

We know there are 1021 time steps per SSP (1021 months).

In [49]:
ssp_train = ['ssp126', 'ssp585']
ssp_val = 'ssp370'
ssp_test = 'ssp245'

n_members = ds.sizes['member_id']
n_time_total = ds.sizes['time']
# print(n_time_total)
lat = ds.sizes['latitude']
lon = ds.sizes['longitude']

# TRAINING = full ssp126 + ssp585 + first 901 months of ssp370
# (since last 10 years / 120 months of SSP370 is the validation set)
n_train_ssp126 = n_time_total * n_members
n_train_ssp585 = n_time_total * n_members
n_train_ssp370 = (n_time_total - 120) * n_members  # time 0 to 900 inclusive
n_train_samples = n_train_ssp126 + n_train_ssp585 + n_train_ssp370

# VALIDATION = last 120 months of ssp370, member_id = 0
n_val_samples = 120

# TEST = last 360 months of ssp245, member_id = 0
n_test_samples = 360

print(f"Spatial Dimensions per sample: (lat, lon) =, {(lat, lon)} --> 3456 grid points")
print("Training samples:", n_train_samples)
print("Validation samples:", n_val_samples)
print("Test samples:", n_test_samples)
n_train_samples + n_val_samples + n_test_samples

Spatial Dimensions per sample: (lat, lon) =, (48, 72) --> 3456 grid points
Training samples: 8829
Validation samples: 120
Test samples: 360


9309

In [50]:
lat_dim = ds.sizes["latitude"]
lon_dim = ds.sizes["longitude"]
n_spatial = lat_dim * lon_dim
print(f"Spatial dimensions: {lat_dim} x {lon_dim} = {n_spatial} grid points")

Spatial dimensions: 48 x 72 = 3456 grid points


**1b)** The dataset consists of 8,829 training samples drawn from ssp126, ssp370 (first 901 months), and ssp585 across all 3 ensemble members. Validation includes 120 samples from the last 10 years (120 months) of ssp370 using only member_id = 0. The test set contains 360 samples from the last 360 months of ssp245, also using only member_id = 0. We only use member_id = 0 because the competition evaluation is performed on a single ensemble realization (member_id = 0), and using multiple ensemble members for validation or testing would misrepresent model performance (data leakage from internal climate variability across ensemble runs).

There are 2 spatial dimensions (latitude and longitude) forming a 48 × 72 global grid. Each sample covers the full grid, so each sample contains 3,456 spatial points.

**Distribution of Target Variables, Input Data**

In [51]:
ds["tas"]
ds["tas"].sel(ssp="ssp245", member_id=0)


<xarray.DataArray 'tas' (time: 1021, y: 48, x: 72)> Size: 14MB
dask.array<getitem, shape=(1021, 48, 72), dtype=float32, chunksize=(24, 48, 72), chunktype=numpy.ndarray>
Coordinates:
    lat        (x, y) float64 28kB dask.array<chunksize=(72, 48), meta=np.ndarray>
    lon        (x, y) float64 28kB dask.array<chunksize=(72, 48), meta=np.ndarray>
    member_id  int64 8B 0
    ssp        <U6 24B 'ssp245'
  * time       (time) object 8kB 2015-01-15 00:00:00 ... 2100-01-15 00:00:00
  * x          (x) float64 576B 1.875 6.875 11.88 16.88 ... 346.9 351.9 356.9
  * y          (y) float64 384B -88.59 -84.82 -81.05 ... 81.05 84.82 88.59
Attributes: (12/19)
    cell_measures:  area: areacella
    cell_methods:   area: time: mean
    comment:        TREFHT
    description:    near-surface (usually, 2 meter) air temperature
    frequency:      day
    id:             tas
    ...             ...
    time_label:     time-mean
    time_title:     Temporal mean
    title:          Near-Surface Air Temperature
    type:           real
    units:          K
    variable_id:    tas

In [52]:
ds["tas"].sel(ssp="ssp245", member_id=0).isel(time=0)


<xarray.DataArray 'tas' (y: 48, x: 72)> Size: 14kB
dask.array<getitem, shape=(48, 72), dtype=float32, chunksize=(48, 72), chunktype=numpy.ndarray>
Coordinates:
    lat        (x, y) float64 28kB dask.array<chunksize=(72, 48), meta=np.ndarray>
    lon        (x, y) float64 28kB dask.array<chunksize=(72, 48), meta=np.ndarray>
    member_id  int64 8B 0
    ssp        <U6 24B 'ssp245'
    time       object 8B 2015-01-15 00:00:00
  * x          (x) float64 576B 1.875 6.875 11.88 16.88 ... 346.9 351.9 356.9
  * y          (y) float64 384B -88.59 -84.82 -81.05 ... 81.05 84.82 88.59
Attributes: (12/19)
    cell_measures:  area: areacella
    cell_methods:   area: time: mean
    comment:        TREFHT
    description:    near-surface (usually, 2 meter) air temperature
    frequency:      day
    id:             tas
    ...             ...
    time_label:     time-mean
    time_title:     Temporal mean
    title:          Near-Surface Air Temperature
    type:           real
    units:          K
    variable_id:    tas

In [53]:
ssp_train = ["ssp126", "ssp370", "ssp585"]
input_vars = ["CO2", "CH4", "SO2", "BC", "rsdt"]
output_vars = ["tas", "pr"]

def clean_flat(arr):
    return arr.compute().values.flatten()[~np.isnan(arr.compute().values.flatten())]

def print_stats(values):
    zero_count = (values == 0).sum()
    print(
        f"count = {len(values):>6} | "
        f"mean = {np.mean(values)} | "
        f"std = {np.std(values)} | "
        f"median = {np.median(values)} | "
        f"min = {np.min(values)} | "
        f"max = {np.max(values)} | "
        f"5th–95th pct = {np.percentile(values, 5)} – {np.percentile(values, 95)} | "
        f"zeros = {zero_count} ({100 * zero_count / len(values)}%)"
    )

# input data
for var in input_vars:
    print(f"\n--------- INPUT VARIABLE: {var} ---------")
    for ssp in ssp_train:
        print(f"SSP: {ssp}")
        arr = ds[var].sel(ssp=ssp)
        if "member_id" in arr.dims:
            arr = arr.sel(member_id=0)
        values = clean_flat(arr)
        
        print_stats(values)

        plt.hist(values, bins=100)
        plt.title(f"{var} Distribution – {ssp}")
        plt.xlabel(var)
        plt.ylabel("Frequency")
        plt.grid(True)
        plt.tight_layout()
        plt.savefig(f"figures/{var}_{ssp}_dist.png", dpi=300)
        plt.close()
        # plt.show()

# target vars
for var in output_vars:
    print(f"\n--------- TARGET VARIABLE: {var} ---------")
    for ssp in ssp_train:
        print(f"SSP: {ssp}")
        arr = ds[var].sel(ssp=ssp, member_id=0)
        values = clean_flat(arr)

        print_stats(values)

        plt.hist(values, bins=100)
        plt.title(f"{var} Distribution – {ssp}")
        plt.xlabel(var)
        plt.ylabel("Frequency")
        plt.grid(True)
        plt.tight_layout()
        plt.savefig(f"figures/{var}_{ssp}_dist.png", dpi=300)
        plt.close()
        # plt.show()


--------- INPUT VARIABLE: CO2 ---------
SSP: ssp126
count =   1021 | mean = 2532.318713977056 | std = 401.69992064797714 | median = 2731.6626772312566 | min = 1536.0722224547292 | max = 2891.7034108052826 | 5th–95th pct = 1689.6271652350188 – 2889.948210871908 | zeros = 0 (0.0%)
SSP: ssp370
count =   1021 | mean = 3874.4328704822187 | std = 1528.9858993886194 | median = 3751.897249353455 | min = 1536.0722224547292 | max = 6763.858968521363 | 5th–95th pct = 1699.3051121573612 – 6426.770585143383 | zeros = 0 (0.0%)
SSP: ssp585
count =   1021 | mean = 4653.006907184912 | std = 2343.882160473454 | median = 4182.662019889618 | min = 1536.0722224547292 | max = 9362.59353129993 | 5th–95th pct = 1696.662145353775 – 8816.81862612959 | zeros = 0 (0.0%)

--------- INPUT VARIABLE: CH4 ---------
SSP: ssp126
count =   1021 | mean = 0.1990533398174867 | std = 0.07149415080379436 | median = 0.18533224375025475 | min = 0.11364517919757215 | max = 0.37373672410518455 | 5th–95th pct = 0.1164790331243431

In [54]:
def plot_distribution_across_ssps(ds, var, time_idx, ssp_list=["ssp126", "ssp370", "ssp585"], member_id=0, bins=100):
    os.makedirs("figures", exist_ok=True)
    plt.figure(figsize=(8, 8))

    for ssp in ssp_list:
        arr = ds[var].sel(ssp=ssp, member_id=member_id).isel(time=time_idx)
        values = arr.compute().values.flatten()
        values = values[~np.isnan(values)]

        plt.hist(values, bins=bins, alpha=0.3, label=ssp)

    try:
        time_str = str(ds["time"].isel(time=time_idx).values)[:10].replace("-", "")
    except:
        time_str = f"t{time_idx}"

    plt.title(f"Distribution of {var} at {time_str} Across SSPs")
    plt.xlabel(f"{var} value")
    plt.ylabel("Frequency")
    plt.legend()
    plt.grid(True)
    plt.tight_layout()

    # Save instead of showing
    filename = f"figures/{var}_ssp_comparison_{time_str}.png"
    plt.savefig(filename, dpi=300)
    plt.close()
    print(f"Saved: {filename}")
    # plt.show()


In [55]:
plot_distribution_across_ssps(ds, var="tas", time_idx=0)
plot_distribution_across_ssps(ds, var="pr", time_idx=0)
plot_distribution_across_ssps(ds, var="tas", time_idx=1020)
plot_distribution_across_ssps(ds, var="pr", time_idx=1020)

Saved: figures/tas_ssp_comparison_20150115.png
Saved: figures/pr_ssp_comparison_20150115.png
Saved: figures/tas_ssp_comparison_21000115.png
Saved: figures/pr_ssp_comparison_21000115.png


In [56]:
def plot_distribution_over_time(ds, ssp, var, time_indices, member_id=0, bins=100):
    os.makedirs("figures", exist_ok=True)
    plt.figure(figsize=(8, 5))

    label_parts = []

    for time_idx in time_indices:
        arr = ds[var].sel(ssp=ssp, member_id=member_id).isel(time=time_idx)
        values = arr.compute().values.flatten()
        values = values[~np.isnan(values)]

        try:
            time_label = str(ds["time"].isel(time=time_idx).values)[:10]
            label_parts.append(time_label.replace("-", ""))
        except:
            time_label = f"t{time_idx}"
            label_parts.append(time_label)

        plt.hist(values, bins=bins, alpha=0.5, label=time_label)

    label_str = "_".join(label_parts)
    plt.title(f"{var.upper()} Distribution Over Time – {ssp}")
    plt.xlabel(f"{var} value")
    plt.ylabel("Frequency")
    plt.legend()
    plt.grid(True)
    plt.tight_layout()

    filename = f"figures/{var}_{ssp}_overtime_{label_str}.png"
    plt.savefig(filename, dpi=300)
    plt.close()
    print(f"Saved: {filename}")
    # plt.show()


In [57]:
plot_distribution_over_time(ds, "ssp126", "tas", [0, 1020])
plot_distribution_over_time(ds, "ssp126", "pr", [0, 1020])
plot_distribution_over_time(ds, "ssp585", "tas", [0, 1020])
plot_distribution_over_time(ds, "ssp585", "pr", [0, 1020])

Saved: figures/tas_ssp126_overtime_20150115_21000115.png
Saved: figures/pr_ssp126_overtime_20150115_21000115.png
Saved: figures/tas_ssp585_overtime_20150115_21000115.png
Saved: figures/pr_ssp585_overtime_20150115_21000115.png


### 🔧 Spatial Weighting Utility Function

This cell sets up utility functions for reproducibility and spatial weighting:

- **`get_lat_weights(latitude_values)`**: Computes cosine-based area weights for each latitude, accounting for the Earth's curvature. This is critical for evaluating global climate metrics fairly — grid cells near the equator represent larger surface areas than those near the poles.


In [58]:
def get_lat_weights(latitude_values):
    lat_rad = np.deg2rad(latitude_values)
    weights = np.cos(lat_rad)
    return weights / np.mean(weights)

### 🧠 SimpleCNN: A Residual Convolutional Baseline

This is a lightweight baseline model designed to capture spatial patterns in global climate data using convolutional layers.

- The architecture starts with a **convolution + batch norm + ReLU** block to process the input channels.
- It then applies a series of **residual blocks** to extract increasingly abstract spatial features. These help preserve gradient flow during training.
- Finally, a few convolutional layers reduce the feature maps down to the desired number of output channels (`tas` and `pr`).

This model only serves as a **simple baseline for climate emulation**. 

We encourage you to build and experiment with your own models and ideas.


In [59]:
# class ResidualBlock(nn.Module):
#     def __init__(self, in_channels, out_channels, kernel_size=3, stride=1):
#         super().__init__()
#         self.conv1 = nn.Conv2d(in_channels, out_channels, kernel_size, stride=stride, padding=kernel_size // 2)
#         self.bn1 = nn.BatchNorm2d(out_channels)
#         self.relu = nn.ReLU(inplace=True)
#         self.conv2 = nn.Conv2d(out_channels, out_channels, kernel_size, padding=kernel_size // 2)
#         self.bn2 = nn.BatchNorm2d(out_channels)
#         self.skip = nn.Sequential()
#         if stride != 1 or in_channels != out_channels:
#             self.skip = nn.Sequential(
#                 nn.Conv2d(in_channels, out_channels, kernel_size=1, stride=stride), nn.BatchNorm2d(out_channels)
#             )

#     def forward(self, x):
#         identity = x
#         out = self.relu(self.bn1(self.conv1(x)))
#         out = self.bn2(self.conv2(out))
#         out += self.skip(identity)
#         return self.relu(out)

# class SimpleCNN(nn.Module):
#     def __init__(self, n_input_channels, n_output_channels, kernel_size=3, init_dim=64, depth=4, dropout_rate=0.2):
#         super().__init__()
#         self.initial = nn.Sequential(
#             nn.Conv2d(n_input_channels, init_dim, kernel_size=kernel_size, padding=kernel_size // 2),
#             nn.BatchNorm2d(init_dim),
#             nn.ReLU(inplace=True),
#         )
#         self.res_blocks = nn.ModuleList()
#         current_dim = init_dim
#         for i in range(depth):
#             out_dim = current_dim * 2 if i < depth - 1 else current_dim
#             self.res_blocks.append(ResidualBlock(current_dim, out_dim))
#             if i < depth - 1:
#                 current_dim *= 2
#         self.dropout = nn.Dropout2d(dropout_rate)
#         self.final = nn.Sequential(
#             nn.Conv2d(current_dim, current_dim // 2, kernel_size=kernel_size, padding=kernel_size // 2),
#             nn.BatchNorm2d(current_dim // 2),
#             nn.ReLU(inplace=True),
#             nn.Conv2d(current_dim // 2, n_output_channels, kernel_size=1),
#         )

#     def forward(self, x):
#         x = self.initial(x)
#         for res_block in self.res_blocks:
#             x = res_block(x)
#         return self.final(self.dropout(x))


### ConvLSTM

In [60]:
# import torch
# import torch.nn as nn
# from torch.nn import Conv2d, BatchNorm2d, ReLU, Dropout2d

# class ConvLSTMCell(nn.Module):
#     def __init__(self, input_dim, hidden_dim, kernel_size):
#         super().__init__()
#         padding = kernel_size // 2
#         self.conv = nn.Conv2d(input_dim + hidden_dim, 4 * hidden_dim, kernel_size, padding=padding)

#     def forward(self, x, h_prev, c_prev):
#         combined = torch.cat([x, h_prev], dim=1)  # (B, C+H, H, W)
#         conv_output = self.conv(combined)
#         cc_i, cc_f, cc_o, cc_g = torch.split(conv_output, conv_output.size(1) // 4, dim=1)
#         i = torch.sigmoid(cc_i)
#         f = torch.sigmoid(cc_f)
#         o = torch.sigmoid(cc_o)
#         g = torch.tanh(cc_g)
#         c_next = f * c_prev + i * g
#         h_next = o * torch.tanh(c_next)
#         return h_next, c_next

# class ConvLSTM(nn.Module):
#     def __init__(self, input_dim, hidden_dim, kernel_size, num_layers=1):
#         super().__init__()
#         self.layers = nn.ModuleList([
#             ConvLSTMCell(input_dim if i == 0 else hidden_dim, hidden_dim, kernel_size)
#             for i in range(num_layers)
#         ])

#     def forward(self, x):
#         # x shape: (B, T, C, H, W)
#         B, T, C, H, W = x.size()
#         h = [torch.zeros(B, layer.conv.out_channels // 4, H, W, device=x.device) for layer in self.layers]
#         c = [torch.zeros(B, layer.conv.out_channels // 4, H, W, device=x.device) for layer in self.layers]

#         for t in range(T):
#             input_ = x[:, t]
#             for i, layer in enumerate(self.layers):
#                 h[i], c[i] = layer(input_, h[i], c[i])
#                 input_ = h[i]

#         return h[-1]  # final hidden state of last layer

# class ConvLSTMModel(nn.Module):
#     def __init__(self, n_input_channels, n_output_channels, T=3, kernel_size=3, hidden_dim=64, dropout_rate=0.1):
#         super().__init__()
#         self.T = T  # ✅ dynamic T
#         self.C = n_input_channels // T
#         self.hidden_dim = hidden_dim

#         self.convlstm = ConvLSTM(self.C, hidden_dim, kernel_size, num_layers=2)  # or configurable

#         self.decoder = nn.Sequential(
#             Conv2d(hidden_dim, hidden_dim, kernel_size=3, padding=1),
#             BatchNorm2d(hidden_dim),
#             ReLU(inplace=True),
#             Dropout2d(dropout_rate),
#             Conv2d(hidden_dim, hidden_dim // 2, kernel_size=3, padding=1),
#             BatchNorm2d(hidden_dim // 2),
#             ReLU(inplace=True),
#             Dropout2d(dropout_rate),
#             Conv2d(hidden_dim // 2, n_output_channels, kernel_size=1)
#         )

#     def forward(self, x):
#         B, C, H, W = x.shape
#         x = x.view(B, self.T, self.C, H, W)  # reshape to (B, T, C, H, W)
#         x = self.convlstm(x)                # output shape: (B, hidden_dim, H, W)
#         return self.decoder(x)              # output shape: (B, 2, H, W)


In [61]:
import torch
import torch.nn as nn
from torch.nn import Conv2d, BatchNorm2d, ReLU, Dropout2d
import torch.nn.functional as F


class ConvLSTMCell(nn.Module):
    def __init__(self, input_dim, hidden_dim, kernel_size):
        super().__init__()
        padding = kernel_size // 2
        self.conv = nn.Conv2d(input_dim + hidden_dim, 4 * hidden_dim, kernel_size, padding=padding)

    def forward(self, x, h_prev, c_prev):
        combined = torch.cat([x, h_prev], dim=1)
        conv_output = self.conv(combined)
        cc_i, cc_f, cc_o, cc_g = torch.chunk(conv_output, 4, dim=1)
        i = torch.sigmoid(cc_i)
        f = torch.sigmoid(cc_f)
        o = torch.sigmoid(cc_o)
        g = torch.tanh(cc_g)
        c_next = f * c_prev + i * g
        h_next = o * torch.tanh(c_next)
        return h_next, c_next


class ConvLSTM(nn.Module):
    def __init__(self, input_dim, hidden_dim, kernel_size, num_layers=1):
        super().__init__()
        self.layers = nn.ModuleList([
            ConvLSTMCell(input_dim if i == 0 else hidden_dim, hidden_dim, kernel_size)
            for i in range(num_layers)
        ])

    def forward(self, x):
        B, T, C, H, W = x.size()
        h = [torch.zeros(B, layer.conv.out_channels // 4, H, W, device=x.device) for layer in self.layers]
        c = [torch.zeros(B, layer.conv.out_channels // 4, H, W, device=x.device) for layer in self.layers]

        skip_connections = []

        for t in range(T):
            input_ = x[:, t]
            for i, layer in enumerate(self.layers):
                h[i], c[i] = layer(input_, h[i], c[i])
                input_ = h[i]
            if t == T - 1:
                skip_connections.append(input_)  # final hidden state for skip

        return h[-1], skip_connections  # final output and skip connection


class UNetDecoderBlock(nn.Module):
    def __init__(self, up_channels, skip_channels, out_channels, dropout_rate=0.1):
        super().__init__()
        self.up = nn.ConvTranspose2d(up_channels, out_channels, kernel_size=2, stride=2)
        self.conv = nn.Sequential(
            Conv2d(out_channels + skip_channels, out_channels, kernel_size=3, padding=1),
            BatchNorm2d(out_channels),
            ReLU(inplace=True),
            Dropout2d(dropout_rate),
            Conv2d(out_channels, out_channels, kernel_size=3, padding=1),
            BatchNorm2d(out_channels),
            ReLU(inplace=True)
        )

    def forward(self, x, skip):
        x = self.up(x)
        if skip.shape[2:] != x.shape[2:]:
            skip = F.interpolate(skip, size=x.shape[2:], mode="bilinear", align_corners=False)
        x = torch.cat([x, skip], dim=1)
        return self.conv(x)



class ConvLSTMUNet(nn.Module):
    def __init__(self, n_input_channels, n_output_channels, T, kernel_size=3, hidden_dim=64, dropout_rate=0.1):
        super().__init__()
        self.T = T
        self.C = n_input_channels // T

        self.convlstm = ConvLSTM(self.C, hidden_dim, kernel_size, num_layers=2)

        self.dec1 = UNetDecoderBlock(
            up_channels=hidden_dim,
            skip_channels=hidden_dim,
            out_channels=hidden_dim // 2,
            dropout_rate=dropout_rate
        )

        self.final = nn.Sequential(
            nn.Conv2d(hidden_dim // 2, n_output_channels, kernel_size=1)
        )


    def forward(self, x):
        B, C, H, W = x.shape
        x = x.view(B, self.T, self.C, H, W)
        x, skips = self.convlstm(x)
        x = self.dec1(x, skips[-1])
        x = self.final(x)
        x = F.interpolate(x, size=(48, 72), mode="bilinear", align_corners=False)
        return x

### 📐 Normalizer: Z-Score Scaling for Climate Inputs & Outputs

This class handles **Z-score normalization**, a crucial preprocessing step for stable and efficient neural network training:

- **`set_input_statistics(mean, std)` / `set_output_statistics(...)`**: Store the mean and standard deviation computed from the training data for later use.
- **`normalize(data, data_type)`**: Standardizes the data using `(x - mean) / std`. This is applied separately to inputs and outputs.
- **`inverse_transform_output(data)`**: Converts model predictions back to the original physical units (e.g., Kelvin for temperature, mm/day for precipitation).

Normalizing the data ensures the model sees inputs with similar dynamic ranges and avoids biases caused by different variable scales.


In [62]:
class Normalizer:
    def __init__(self):
        self.mean_in, self.std_in = None, None
        self.mean_out, self.std_out = None, None

    def set_input_statistics(self, mean, std):
        self.mean_in = mean
        self.std_in = std

    def set_output_statistics(self, mean, std):
        self.mean_out = mean
        self.std_out = std

    def normalize(self, data, data_type):
        if data_type == "input":
            return (data - self.mean_in) / self.std_in
        elif data_type == "output":
            return (data - self.mean_out) / self.std_out

    def inverse_transform_output(self, data):
        return data * self.std_out + self.mean_out


### 🌍 Data Module: Loading, Normalization, and Splitting

This section handles the entire data pipeline, from loading the `.zarr` dataset to preparing PyTorch-ready DataLoaders.

#### `ClimateDataset`
- A simple PyTorch `Dataset` wrapper that preloads the entire (normalized) dataset into memory using Dask.
- Converts the data to PyTorch tensors and handles any `NaN` checks up front.

#### `ClimateDataModule`
A PyTorch Lightning `DataModule` that handles:
- ✅ **Loading data** from different SSP scenarios and ensemble members
- ✅ **Broadcasting non-spatial inputs** (like CO₂) to match spatial grid size
- ✅ **Normalization** using mean/std computed from training data only
- ✅ **Splitting** into training, validation, and test sets:
  - Training: All months from selected SSPs (except last 10 years of SSP370)
  - Validation: Last 10 years (120 months) of SSP370
  - Test: Last 10 years of SSP245 (unseen scenario)
- ✅ **Batching** and parallelized data loading via PyTorch `DataLoader`s
- ✅ **Latitude-based area weighting** for fair climate metric evaluation
- Shape of the inputs are Batch_Size X 5 (num_input_variables) X 48 X 72
- Shape of ouputputs are Batch_Size X 2 (num_output_variables) X 48 X 72

> ℹ️ **Note:** You likely won’t need to modify this class but feel free to make modifications if you want to inlcude different ensemble mebers to feed more data to your models


In [63]:
class ClimateDataset(Dataset):
    # def __init__(self, inputs_dask, outputs_dask, output_is_normalized=True):
    #     self.size = inputs_dask.shape[0]
    #     print(f"Creating dataset with {self.size} samples...")

    #     inputs_np = inputs_dask.compute()
    #     outputs_np = outputs_dask.compute()

    #     self.inputs = torch.from_numpy(inputs_np).float()
    #     self.outputs = torch.from_numpy(outputs_np).float()

    #     if torch.isnan(self.inputs).any() or torch.isnan(self.outputs).any():
    #         raise ValueError("NaNs found in dataset")

    # def __len__(self):
    #     return self.size

    # def __getitem__(self, idx):
    #     return self.inputs[idx], self.outputs[idx]
    def __init__(self, inputs_dask, outputs_dask, T=6, output_is_normalized=True):
        self.inputs_np = inputs_dask.compute()
        self.outputs_np = outputs_dask.compute()
        self.T = T
        self.output_is_normalized = output_is_normalized

        self.size = self.inputs_np.shape[0] - (T - 1)
        print(f"Creating dataset with {self.size} samples (T={T})")

        if np.isnan(self.inputs_np).any() or np.isnan(self.outputs_np).any():
            raise ValueError("NaNs found in dataset")

    def __len__(self):
        return self.size

    def __getitem__(self, idx):
        # Stack T timesteps as channels: shape = (T × input_channels, H, W)
        x = torch.from_numpy(self.inputs_np[idx:idx + self.T]).float()
        x = x.view(-1, *x.shape[2:])  # flatten across time into channels

        y = torch.from_numpy(self.outputs_np[idx + self.T - 1]).float()
        return x, y


class ClimateDataModule(pl.LightningDataModule):
    def __init__(
        self,
        path,
        input_vars,
        output_vars,
        train_ssps,
        test_ssp,
        target_member_id,
        val_split=0.1,
        test_months=360,
        batch_size=32,
        num_workers=0,
        seed=42,
        T=12
    ):
        super().__init__()
        self.path = path
        self.input_vars = input_vars
        self.output_vars = output_vars
        self.train_ssps = train_ssps
        self.test_ssp = test_ssp
        self.target_member_id = target_member_id
        self.val_split = val_split
        self.test_months = test_months
        self.batch_size = batch_size
        self.num_workers = num_workers
        self.seed = seed
        self.normalizer = Normalizer()
        self.T = T

    def prepare_data(self):
        assert os.path.exists(self.path), f"Data path not found: {self.path}"

    def setup(self, stage=None):
        ds = xr.open_zarr(self.path, consolidated=False, chunks={"time": 24})
        spatial_template = ds["rsdt"].isel(time=0, ssp=0, drop=True)

        def load_ssp(ssp):
            input_dask, output_dask = [], []
            for var in self.input_vars:
                da_var = ds[var].sel(ssp=ssp)
                if "latitude" in da_var.dims:
                    da_var = da_var.rename({"latitude": "y", "longitude": "x"})
                if "member_id" in da_var.dims:
                    da_var = da_var.sel(member_id=self.target_member_id)
                if set(da_var.dims) == {"time"}:
                    da_var = da_var.broadcast_like(spatial_template).transpose("time", "y", "x")
                input_dask.append(da_var.data)

            for var in self.output_vars:
                da_out = ds[var].sel(ssp=ssp, member_id=self.target_member_id)
                if "latitude" in da_out.dims:
                    da_out = da_out.rename({"latitude": "y", "longitude": "x"})
                output_dask.append(da_out.data)

            return da.stack(input_dask, axis=1), da.stack(output_dask, axis=1)

        train_input, train_output, val_input, val_output = [], [], None, None

        for ssp in self.train_ssps:
            x, y = load_ssp(ssp)
            if ssp == "ssp370":
                val_input = x[-self.test_months:]
                val_output = y[-self.test_months:]
                train_input.append(x[:-self.test_months])
                train_output.append(y[:-self.test_months])
            else:
                train_input.append(x)
                train_output.append(y)

        train_input = da.concatenate(train_input, axis=0)
        train_output = da.concatenate(train_output, axis=0)

        self.normalizer.set_input_statistics(
            mean=da.nanmean(train_input, axis=(0, 2, 3), keepdims=True).compute(),
            std=da.nanstd(train_input, axis=(0, 2, 3), keepdims=True).compute(),
        )
        self.normalizer.set_output_statistics(
            mean=da.nanmean(train_output, axis=(0, 2, 3), keepdims=True).compute(),
            std=da.nanstd(train_output, axis=(0, 2, 3), keepdims=True).compute(),
        )

        train_input_norm = self.normalizer.normalize(train_input, "input")
        train_output_norm = self.normalizer.normalize(train_output, "output")
        val_input_norm = self.normalizer.normalize(val_input, "input")
        val_output_norm = self.normalizer.normalize(val_output, "output")

        test_input, test_output = load_ssp(self.test_ssp)
        # test_input = test_input[-self.test_months:]
        # test_output = test_output[-self.test_months:]

        # Replace your current two lines with these
        test_input  = test_input[-(self.test_months + self.T - 1):]   # 360 + 11 = 371 frames
        test_output = test_output[-(self.test_months + self.T - 1):]  # 371 frames

        test_input_norm = self.normalizer.normalize(test_input, "input")

        # self.train_dataset = ClimateDataset(train_input_norm, train_output_norm)
        # self.val_dataset = ClimateDataset(val_input_norm, val_output_norm)
        # self.test_dataset = ClimateDataset(test_input_norm, test_output, output_is_normalized=False)
        self.train_dataset = ClimateDataset(train_input_norm, train_output_norm, T=self.T)
        self.val_dataset = ClimateDataset(val_input_norm, val_output_norm, T=self.T)
        # self.test_dataset = ClimateDataset(test_input_norm, test_output, T=self.T, output_is_normalized=False)
        # FIXED TEST DATASET
        self.test_dataset = ClimateDataset(
            test_input_norm,
            test_output,
            T=self.T,
            output_is_normalized=False
        )

        self.lat = spatial_template.y.values
        self.lon = spatial_template.x.values
        self.area_weights = xr.DataArray(get_lat_weights(self.lat), dims=["y"], coords={"y": self.lat})

    def train_dataloader(self):
        return DataLoader(self.train_dataset, batch_size=self.batch_size, shuffle=True,
                          num_workers=self.num_workers, pin_memory=True)

    def val_dataloader(self):
        return DataLoader(self.val_dataset, batch_size=self.batch_size, shuffle=False,
                          num_workers=self.num_workers, pin_memory=True)

    def test_dataloader(self):
        return DataLoader(self.test_dataset, batch_size=self.batch_size, shuffle=False,
                          num_workers=self.num_workers, pin_memory=True)

    def get_lat_weights(self):
        return self.area_weights

    def get_coords(self):
        return self.lat, self.lon

### ⚡ ClimateEmulationModule: Lightning Wrapper for Climate Model Emulation

This is the core model wrapper built with **PyTorch Lightning**, which organizes the training, validation, and testing logic for the climate emulation task. Lightning abstracts away much of the boilerplate code in PyTorch-based deep learning workflows, making it easier to scale models.

#### ✅ Key Features

- **`training_step` / `validation_step` / `test_step`**: Standard Lightning hooks for computing loss and predictions at each stage. The loss used is **Mean Squared Error (MSE)**.

- **Normalization-aware outputs**:
  - During validation and testing, predictions and targets are denormalized before evaluation using stored mean/std statistics.
  - This ensures evaluation is done in real-world units (Kelvin and mm/day).

- **Metric Evaluation** via `_evaluate()`:
  For each variable (`tas`, `pr`), it calculates:
  - **Monthly Area-Weighted RMSE**
  - **Time-Mean RMSE** (RMSE on 10-year average's)
  - **Time-Stddev MAE** (MAE on 10-year standard deviation; a measure of temporal variability)
    
  These metrics reflect the competition's evaluation criteria and are logged and printed.

- **Kaggle Submission Writer**:
  After testing, predictions are saved to a `.csv` file in the required Kaggle format via `_save_submission()`.

- **Saving Predictions for Visualization**:
  - Validation predictions are saved tao `val_preds.npy` and `val_trues.npy`
  - These can be loaded later for visual inspection of the model's performance.

 🔧 **Feel free to modify any part of this module** (loss functions, evaluation, training logic) to better suit your model or training pipeline / Use pure PyTorch etc.

⚠️ The **final submission `.csv` file must strictly follow the format and naming convention used in `_save_submission()`**, as these `ID`s are used to match predictions to the hidden test set during evaluation.



In [64]:
import pandas as pd
from torch.optim.lr_scheduler import ReduceLROnPlateau

class ClimateEmulationModule(pl.LightningModule):
    def __init__(self, model, learning_rate=1e-4):
        super().__init__()
        self.model = model
        self.save_hyperparameters(ignore=['model']) # Save all hyperparameters except the model to self.hparams.<param_name>
        self.criterion = nn.MSELoss()
        self.normalizer = None
        self.val_preds, self.val_targets = [], []
        self.test_preds, self.test_targets = [], []

    def forward(self, x):
        return self.model(x)

    def on_fit_start(self):
        self.normalizer = self.trainer.datamodule.normalizer  # Get the normalizer from the datamodule (see above)

    def training_step(self, batch, batch_idx):
        x, y = batch # Unpack inputs and targets (this is the output of the _getitem_ method in the Dataset above)
        y_hat = self(x)   # Forward pass
        loss = self.criterion(y_hat, y)  # Calculate loss
        self.log("train/loss", loss)  # Log loss for tracking
        return loss

    def validation_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = self.criterion(y_hat, y)
        self.log("val/loss", loss)

        y_hat_np = self.normalizer.inverse_transform_output(y_hat.detach().cpu().numpy())
        y_np = self.normalizer.inverse_transform_output(y.detach().cpu().numpy())
        self.val_preds.append(y_hat_np)
        self.val_targets.append(y_np)

        return loss

    def on_validation_epoch_end(self):
        # Concatenate all predictions and ground truths from each val step/batch into one array
        preds = np.concatenate(self.val_preds, axis=0)
        trues = np.concatenate(self.val_targets, axis=0)
        self._evaluate(preds, trues, phase="val")
        np.save("val_preds.npy", preds)
        np.save("val_trues.npy", trues)
        self.val_preds.clear()
        self.val_targets.clear()

    def test_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        y_hat_np = self.normalizer.inverse_transform_output(y_hat.detach().cpu().numpy())
        y_np = y.detach().cpu().numpy()
        self.test_preds.append(y_hat_np)
        self.test_targets.append(y_np)

    def on_test_epoch_end(self):
        # Concatenate all predictions and ground truths from each test step/batch into one array
        preds = np.concatenate(self.test_preds, axis=0)
        print("✅ predictions shape:", preds.shape)
        
        trues = np.concatenate(self.test_targets, axis=0)
        self._evaluate(preds, trues, phase="test")
        self._save_submission(preds)
        self.test_preds.clear()
        self.test_targets.clear()

    # def configure_optimizers(self):
    #     return optim.Adam(self.parameters(), lr=self.hparams.learning_rate)

    def configure_optimizers(self):
        optimizer = optim.Adam(self.parameters(), lr=self.hparams.learning_rate)
        scheduler = {
            'scheduler': ReduceLROnPlateau(optimizer, mode='min', patience=3, factor=0.5, verbose=True),
            'monitor': 'val/loss',
            'interval': 'epoch',
            'frequency': 1
        }
        return {'optimizer': optimizer, 'lr_scheduler': scheduler}

    def _evaluate(self, preds, trues, phase="val"):
        datamodule = self.trainer.datamodule
        area_weights = datamodule.get_lat_weights()
        lat, lon = datamodule.get_coords()
        time = np.arange(preds.shape[0])
        output_vars = datamodule.output_vars

        for i, var in enumerate(output_vars):
            p = preds[:, i]
            t = trues[:, i]
            p_xr = xr.DataArray(p, dims=["time", "y", "x"], coords={"time": time, "y": lat, "x": lon})
            t_xr = xr.DataArray(t, dims=["time", "y", "x"], coords={"time": time, "y": lat, "x": lon})

            # RMSE
            rmse = np.sqrt(((p_xr - t_xr) ** 2).weighted(area_weights).mean(("time", "y", "x")).item())
            # RMSE of time-mean
            mean_rmse = np.sqrt(((p_xr.mean("time") - t_xr.mean("time")) ** 2).weighted(area_weights).mean(("y", "x")).item())
            # MAE of time-stddev
            std_mae = np.abs(p_xr.std("time") - t_xr.std("time")).weighted(area_weights).mean(("y", "x")).item()

            print(f"[{phase.upper()}] {var}: RMSE={rmse:.4f}, Time-Mean RMSE={mean_rmse:.4f}, Time-Stddev MAE={std_mae:.4f}")
            self.log_dict({
                f"{phase}/{var}/rmse": rmse,
                f"{phase}/{var}/time_mean_rmse": mean_rmse,
                f"{phase}/{var}/time_std_mae": std_mae,
            })

    # # original
    # def _save_submission(self, predictions):
    #     datamodule = self.trainer.datamodule
    #     lat, lon = datamodule.get_coords()
    #     output_vars = datamodule.output_vars
    #     time = np.arange(predictions.shape[0])

    #     rows = []
    #     for t_idx, t in enumerate(time):
    #         for var_idx, var in enumerate(output_vars):
    #             for y_idx, y in enumerate(lat):
    #                 for x_idx, x in enumerate(lon):
    #                     row_id = f"t{t_idx:03d}_{var}_{y:.2f}_{x:.2f}"
    #                     pred = predictions[t_idx, var_idx, y_idx, x_idx]
    #                     rows.append({"ID": row_id, "Prediction": pred})

    #     df = pd.DataFrame(rows)
    #     os.makedirs("submissions", exist_ok=True)
    #     filepath = f"submissions/kaggle_submission_{datetime.now().strftime('%Y%m%d_%H%M%S')}.csv"
    #     df.to_csv(filepath, index=False)
    #     print(f"✅ Submission saved to: {filepath}")


    def _save_submission(self, predictions):
        datamodule = self.trainer.datamodule
        lat, lon = datamodule.get_coords()  # shape (48,), (72,)
        output_vars = datamodule.output_vars  # ["tas", "pr"]
    
        T = datamodule.T  # should be 12
        total_time = predictions.shape[0]  # should be 360
        time_indices = np.arange(T, T + total_time)  # starts at 12 --> IDs t012 … t371
    
        rows = []
        for t_idx, t_actual in enumerate(time_indices):
            for var_idx, var in enumerate(output_vars):
                for y_idx, y in enumerate(lat):
                    for x_idx, x in enumerate(lon):
                        row_id = f"t{t_actual:03d}_{var}_{y:.2f}_{x:.2f}"
                        pred = predictions[t_idx, var_idx, y_idx, x_idx]
                        rows.append({"ID": row_id, "Prediction": pred})
    
        df = pd.DataFrame(rows)
        os.makedirs("submissions", exist_ok=True)
        filepath = f"submissions/kaggle_submission_{datetime.now().strftime('%Y%m%d_%H%M%S')}.csv"
        df.to_csv(filepath, index=False)
        print(f"✅ Submission saved to: {filepath}")
        print(f"✅ Submission shape: {df.shape}")




### ⚡ Training & Evaluation with PyTorch Lightning

This block sets up and runs the training and testing pipeline using **PyTorch Lightning’s `Trainer`**, which abstracts away much of the boilerplate in deep learning workflows.

- **Modular Setup**:
  - `datamodule`: Handles loading, normalization, and batching of climate data.
  - `model`: A convolutional neural network that maps climate forcings to predicted outputs.
  - `lightning_module`: Wraps the model with training/validation/test logic and metric evaluation.

- **Trainer Flexibility**:
  The `Trainer` accepts a wide range of configuration options from `config["trainer"]`, including:
  - Number of epochs
  - Precision (e.g., 16-bit or 32-bit)
  - Device configuration (CPU, GPU, or TPU)
  - Determinism, logging, callbacks, and more

In [65]:
# datamodule = ClimateDataModule(**config["data"])
# model = SimpleCNN(
#     n_input_channels=len(config["data"]["input_vars"]),
#     n_output_channels=len(config["data"]["output_vars"]),
#     **{k: v for k, v in config["model"].items() if k != "type"}
# )
# lightning_module = ClimateEmulationModule(model, learning_rate=config["training"]["lr"])

# trainer = pl.Trainer(**config["trainer"])
# trainer.fit(lightning_module, datamodule=datamodule)   # Training

In [68]:
datamodule = ClimateDataModule(**config["data"])

model = ConvLSTMUNet(
    n_input_channels=len(config["data"]["input_vars"]) * config["data"]["T"],
    n_output_channels=len(config["data"]["output_vars"]),
    T=config["data"]["T"],
    kernel_size=config["model"]["kernel_size"],
    hidden_dim=config["model"]["init_dim"],
    dropout_rate=config["model"]["dropout_rate"],
)

print("Model Parameters:")
print(f"  Input Channels     : {len(config['data']['input_vars']) * config['data']['T']}")
print(f"  Output Channels    : {len(config['data']['output_vars'])}")
print(f"  T (timesteps)      : {config['data']['T']}")
print(f"  Kernel Size        : {config['model']['kernel_size']}")
print(f"  Hidden Dim         : {config['model']['init_dim']}")
print(f"  Dropout Rate       : {config['model']['dropout_rate']}")
print(f"  Batch Size       : {config['data']['batch_size']}")

lightning_module = ClimateEmulationModule(model, learning_rate=config["training"]["lr"])

trainer = pl.Trainer(**config["trainer"])
trainer.fit(lightning_module, datamodule=datamodule)

Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs


Model Parameters:
  Input Channels     : 45
  Output Channels    : 2
  T (timesteps)      : 9
  Kernel Size        : 3
  Hidden Dim         : 64
  Dropout Rate       : 0.1
  Batch Size       : 32
Creating dataset with 2695 samples (T=9)
Creating dataset with 352 samples (T=9)


LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name      | Type         | Params | Mode 
---------------------------------------------------
0 | model     | ConvLSTMUNet | 499 K  | train
1 | criterion | MSELoss      | 0      | train
---------------------------------------------------
499 K     Trainable params
0         Non-trainable params
499 K     Total params
1.999     Total estimated model params size (MB)
20        Modules in train mode
0         Modules in eval mode


Creating dataset with 360 samples (T=9)


Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=4.4278, Time-Mean RMSE=3.5308, Time-Stddev MAE=1.3091
[VAL] pr: RMSE=2.9548, Time-Mean RMSE=1.5927, Time-Stddev MAE=1.2808


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=4.0085, Time-Mean RMSE=3.2957, Time-Stddev MAE=1.0717
[VAL] pr: RMSE=2.5150, Time-Mean RMSE=1.1364, Time-Stddev MAE=1.0794


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=4.2328, Time-Mean RMSE=3.5840, Time-Stddev MAE=1.0030
[VAL] pr: RMSE=2.3765, Time-Mean RMSE=1.0176, Time-Stddev MAE=1.0907


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=3.8371, Time-Mean RMSE=3.1899, Time-Stddev MAE=0.9830
[VAL] pr: RMSE=2.2360, Time-Mean RMSE=0.7840, Time-Stddev MAE=0.9900


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=3.7971, Time-Mean RMSE=3.1288, Time-Stddev MAE=1.0124
[VAL] pr: RMSE=2.1419, Time-Mean RMSE=0.6068, Time-Stddev MAE=0.8571


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=3.5318, Time-Mean RMSE=2.8749, Time-Stddev MAE=0.9248
[VAL] pr: RMSE=2.1031, Time-Mean RMSE=0.5733, Time-Stddev MAE=0.8977


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=3.7392, Time-Mean RMSE=3.1792, Time-Stddev MAE=0.8248
[VAL] pr: RMSE=2.1228, Time-Mean RMSE=0.6171, Time-Stddev MAE=0.9310


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=3.3102, Time-Mean RMSE=2.7487, Time-Stddev MAE=0.7215
[VAL] pr: RMSE=2.0701, Time-Mean RMSE=0.4907, Time-Stddev MAE=0.8989


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=3.3994, Time-Mean RMSE=2.8756, Time-Stddev MAE=0.7207
[VAL] pr: RMSE=2.0459, Time-Mean RMSE=0.4497, Time-Stddev MAE=0.8514


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=3.1753, Time-Mean RMSE=2.6367, Time-Stddev MAE=0.6315
[VAL] pr: RMSE=2.0303, Time-Mean RMSE=0.4311, Time-Stddev MAE=0.8630


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=3.2973, Time-Mean RMSE=2.7212, Time-Stddev MAE=0.7354
[VAL] pr: RMSE=2.0755, Time-Mean RMSE=0.5268, Time-Stddev MAE=0.9631


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=3.0910, Time-Mean RMSE=2.5368, Time-Stddev MAE=0.6860
[VAL] pr: RMSE=2.0441, Time-Mean RMSE=0.4710, Time-Stddev MAE=0.8608


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=3.0588, Time-Mean RMSE=2.5530, Time-Stddev MAE=0.6031
[VAL] pr: RMSE=2.0192, Time-Mean RMSE=0.4337, Time-Stddev MAE=0.8733


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=3.0027, Time-Mean RMSE=2.4846, Time-Stddev MAE=0.6178
[VAL] pr: RMSE=2.0282, Time-Mean RMSE=0.4632, Time-Stddev MAE=0.9046


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.9182, Time-Mean RMSE=2.3826, Time-Stddev MAE=0.5823
[VAL] pr: RMSE=2.0196, Time-Mean RMSE=0.4171, Time-Stddev MAE=0.8209


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.7044, Time-Mean RMSE=2.1606, Time-Stddev MAE=0.5191
[VAL] pr: RMSE=2.0172, Time-Mean RMSE=0.4566, Time-Stddev MAE=0.8934


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.9114, Time-Mean RMSE=2.3993, Time-Stddev MAE=0.6151
[VAL] pr: RMSE=2.0039, Time-Mean RMSE=0.3802, Time-Stddev MAE=0.8692


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.8067, Time-Mean RMSE=2.2798, Time-Stddev MAE=0.5863
[VAL] pr: RMSE=2.0025, Time-Mean RMSE=0.4168, Time-Stddev MAE=0.8364


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.9339, Time-Mean RMSE=2.4083, Time-Stddev MAE=0.6427
[VAL] pr: RMSE=2.0120, Time-Mean RMSE=0.4336, Time-Stddev MAE=0.9048


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.5583, Time-Mean RMSE=1.9514, Time-Stddev MAE=0.5289
[VAL] pr: RMSE=1.9938, Time-Mean RMSE=0.4041, Time-Stddev MAE=0.7895


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.7165, Time-Mean RMSE=2.1616, Time-Stddev MAE=0.5263
[VAL] pr: RMSE=2.0054, Time-Mean RMSE=0.4212, Time-Stddev MAE=0.8665


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.6322, Time-Mean RMSE=1.9638, Time-Stddev MAE=0.6035
[VAL] pr: RMSE=2.0001, Time-Mean RMSE=0.4018, Time-Stddev MAE=0.8083


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.3003, Time-Mean RMSE=1.5767, Time-Stddev MAE=0.5417
[VAL] pr: RMSE=2.0230, Time-Mean RMSE=0.4771, Time-Stddev MAE=0.8829


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.3167, Time-Mean RMSE=1.6777, Time-Stddev MAE=0.5414
[VAL] pr: RMSE=2.0000, Time-Mean RMSE=0.3951, Time-Stddev MAE=0.8720


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.3646, Time-Mean RMSE=1.7138, Time-Stddev MAE=0.5820
[VAL] pr: RMSE=2.0188, Time-Mean RMSE=0.4166, Time-Stddev MAE=0.8579


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.2746, Time-Mean RMSE=1.6519, Time-Stddev MAE=0.5354
[VAL] pr: RMSE=1.9866, Time-Mean RMSE=0.3521, Time-Stddev MAE=0.8591


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.0780, Time-Mean RMSE=1.3577, Time-Stddev MAE=0.5344
[VAL] pr: RMSE=1.9919, Time-Mean RMSE=0.3277, Time-Stddev MAE=0.8158


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.3461, Time-Mean RMSE=1.7852, Time-Stddev MAE=0.5127
[VAL] pr: RMSE=1.9916, Time-Mean RMSE=0.3796, Time-Stddev MAE=0.8492


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.0992, Time-Mean RMSE=1.4460, Time-Stddev MAE=0.4626
[VAL] pr: RMSE=2.0134, Time-Mean RMSE=0.4648, Time-Stddev MAE=0.8922


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.1282, Time-Mean RMSE=1.5133, Time-Stddev MAE=0.4737
[VAL] pr: RMSE=1.9824, Time-Mean RMSE=0.3481, Time-Stddev MAE=0.8263


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.0322, Time-Mean RMSE=1.3590, Time-Stddev MAE=0.5104
[VAL] pr: RMSE=1.9928, Time-Mean RMSE=0.4079, Time-Stddev MAE=0.8445


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=1.8918, Time-Mean RMSE=1.2197, Time-Stddev MAE=0.4358
[VAL] pr: RMSE=1.9954, Time-Mean RMSE=0.3659, Time-Stddev MAE=0.8751


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=1.9172, Time-Mean RMSE=1.2143, Time-Stddev MAE=0.4483
[VAL] pr: RMSE=1.9948, Time-Mean RMSE=0.3682, Time-Stddev MAE=0.8363


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=1.8889, Time-Mean RMSE=1.1294, Time-Stddev MAE=0.4431
[VAL] pr: RMSE=1.9745, Time-Mean RMSE=0.3372, Time-Stddev MAE=0.8415


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=1.9544, Time-Mean RMSE=1.2930, Time-Stddev MAE=0.4816
[VAL] pr: RMSE=2.0035, Time-Mean RMSE=0.3556, Time-Stddev MAE=0.9135


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=2.1369, Time-Mean RMSE=1.5676, Time-Stddev MAE=0.4682
[VAL] pr: RMSE=1.9817, Time-Mean RMSE=0.3693, Time-Stddev MAE=0.8631


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=1.9282, Time-Mean RMSE=1.2844, Time-Stddev MAE=0.4336
[VAL] pr: RMSE=1.9889, Time-Mean RMSE=0.3831, Time-Stddev MAE=0.8508


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=1.9206, Time-Mean RMSE=1.2310, Time-Stddev MAE=0.4876
[VAL] pr: RMSE=2.0012, Time-Mean RMSE=0.3945, Time-Stddev MAE=0.9136


Validation: |          | 0/? [00:00<?, ?it/s]

[VAL] tas: RMSE=1.7686, Time-Mean RMSE=1.0953, Time-Stddev MAE=0.4071
[VAL] pr: RMSE=1.9614, Time-Mean RMSE=0.3200, Time-Stddev MAE=0.8202


Validation: |          | 0/? [00:00<?, ?it/s]

`Trainer.fit` stopped: `max_epochs=40` reached.


[VAL] tas: RMSE=1.7863, Time-Mean RMSE=1.0922, Time-Stddev MAE=0.4423
[VAL] pr: RMSE=1.9683, Time-Mean RMSE=0.3332, Time-Stddev MAE=0.8340


# Test model

**IMPORTANT:** Please note that the test metrics will be bad because the test targets have been corrupted on the public Kaggle dataset.
The purpose of testing below is to generate the Kaggle submission file based on your model's predictions, which you can submit to the competition.

In [69]:
trainer.test(lightning_module, datamodule=datamodule) 

Creating dataset with 2695 samples (T=9)
Creating dataset with 352 samples (T=9)


LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]


Creating dataset with 360 samples (T=9)


Testing: |          | 0/? [00:00<?, ?it/s]

✅ predictions shape: (360, 2, 48, 72)
[TEST] tas: RMSE=290.6071, Time-Mean RMSE=290.5709, Time-Stddev MAE=3.3297
[TEST] pr: RMSE=4.0672, Time-Mean RMSE=3.6278, Time-Stddev MAE=1.3110
✅ Submission saved to: submissions/kaggle_submission_20250520_085448.csv
✅ Submission shape: (2488320, 2)


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│       test/pr/rmse        │     4.067189693450928     │
│  test/pr/time_mean_rmse   │     3.627798318862915     │
│   test/pr/time_std_mae    │     1.311017394065857     │
│       test/tas/rmse       │     290.6070861816406     │
│  test/tas/time_mean_rmse  │     290.5708923339844     │
│   test/tas/time_std_mae   │    3.3297204971313477     │
└───────────────────────────┴───────────────────────────┘

[{'test/tas/rmse': 290.6070861816406,
  'test/tas/time_mean_rmse': 290.5708923339844,
  'test/tas/time_std_mae': 3.3297204971313477,
  'test/pr/rmse': 4.067189693450928,
  'test/pr/time_mean_rmse': 3.627798318862915,
  'test/pr/time_std_mae': 1.311017394065857}]

### Plotting Utils


In [ ]:
def plot_comparison(true_xr, pred_xr, title, cmap='viridis', diff_cmap='RdBu_r', metric=None):
    fig, axs = plt.subplots(1, 3, figsize=(18, 6))

    vmin = min(true_xr.min().item(), pred_xr.min().item())
    vmax = max(true_xr.max().item(), pred_xr.max().item())

    # Ground truth
    true_xr.plot(ax=axs[0], cmap=cmap, vmin=vmin, vmax=vmax, add_colorbar=True)
    axs[0].set_title(f"{title} (Ground Truth)")

    # Prediction
    pred_xr.plot(ax=axs[1], cmap=cmap, vmin=vmin, vmax=vmax, add_colorbar=True)
    axs[1].set_title(f"{title} (Prediction)")

    # Difference
    diff = pred_xr - true_xr
    abs_max = np.max(np.abs(diff))
    diff.plot(ax=axs[2], cmap=diff_cmap, vmin=-abs_max, vmax=abs_max, add_colorbar=True)
    axs[2].set_title(f"{title} (Difference) {f'- {metric:.4f}' if metric else ''}")

    plt.tight_layout()
    plt.show()


### 🖼️ Visualizing Validation Predictions

This cell loads saved validation predictions and compares them to the ground truth using spatial plots. These visualizations help you qualitatively assess your model's performance.

For each output variable (`tas`, `pr`), we visualize:

- **📈 Time-Mean Map**: The 10-year average spatial pattern for both prediction and ground truth. Helps identify long-term biases or spatial shifts.
- **📊 Time-Stddev Map**: Shows the standard deviation across time for each grid cell — useful for assessing how well the model captures **temporal variability** at each location.
- **🕓 Random Timestep Sample**: Visual comparison of prediction vs ground truth for a single month. Useful for spotting fine-grained anomalies or errors in specific months.

> These plots provide intuition beyond metrics and are useful for debugging spatial or temporal model failures.


In [ ]:
# Load validation predictions
# make sure to have run the validation loop at least once
val_preds = np.load("val_preds.npy")
val_trues = np.load("val_trues.npy")

lat, lon = datamodule.get_coords()
output_vars = datamodule.output_vars
time = np.arange(val_preds.shape[0])

for i, var in enumerate(output_vars):
    pred_xr = xr.DataArray(val_preds[:, i], dims=["time", "y", "x"], coords={"time": time, "y": lat, "x": lon})
    true_xr = xr.DataArray(val_trues[:, i], dims=["time", "y", "x"], coords={"time": time, "y": lat, "x": lon})

    # --- Time Mean ---
    plot_comparison(true_xr.mean("time"), pred_xr.mean("time"), f"{var} Val Time-Mean")

    # --- Time Stddev ---
    plot_comparison(true_xr.std("time"), pred_xr.std("time"), f"{var} Val Time-Stddev", cmap="plasma")

    # --- Random timestep ---
    t_idx = np.random.randint(0, len(time))
    plot_comparison(true_xr.isel(time=t_idx), pred_xr.isel(time=t_idx), f"{var} Val Sample Timestep {t_idx}")


## 🧪 Final Notes

This notebook is meant to serve as a **baseline template** — a starting point to help you get up and running quickly with the climate emulation challenge.

You are **not** required to stick to this exact setup. In fact, we **encourage** you to:

- 🔁 Build on top of the provided `DataModule`. 
- 🧠 Use your own model architectures or training pipelines that you’re more comfortable with 
- ⚗️ Experiment with ideas  
- 🥇 Compete creatively to climb the Kaggle leaderboard  
- 🙌 Most importantly: **have fun** and **learn as much as you can** along the way

This challenge simulates a real-world scientific problem, and there’s no single "correct" approach — so be curious, experiment boldly, and make it your own!


In [ ]:
# supposed to be 2488320
pd.read_csv('submissions/kaggle_submission_20250519_231518.csv').shape[0]

In [ ]:
range(config["data"]["T"], config["data"]["T"] + config["data"]["test_months"])
